src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -2176,73 +1999,208 @@ lExit: ;
-//      ReflectionSerialization
-//*************************************************************************************************
-//*************************************************************************************************
-//*************************************************************************************************
-FCIMPL1(Object*, ReflectionSerialization::GetUninitializedObject, ReflectClassBaseObject* objTypeUNSAFE) {
-    FCALL_CONTRACT;
+/*
+ * Given a TypeHandle, returns the address of the NEWOBJ helper function that creates
+ * a zero-inited instance of this type. If NEWOBJ is not supported on this TypeHandle,
+ * throws an exception. If TypeHandle is a value type, the NEWOBJ helper will create
+ * a boxed zero-inited instance of the value type. The "fGetUninitializedObject"
+ * parameter dictates whether the caller is RuntimeHelpers.GetUninitializedObject or
+ * Activator.CreateInstance, which have different behavior w.r.t. what exceptions are
+ * thrown on failure and how nullables and COM types are handled.
+ */
+void QCALLTYPE RuntimeTypeHandle::GetAllocatorFtn(
+    QCall::TypeHandle pTypeHandle,
+    PCODE* ppNewobjHelper,
+    MethodTable** ppMT,
+    BOOL fGetUninitializedObject,
+    BOOL* pfFailedWhileRunningCctor)
+{
+    CONTRACTL{
+        QCALL_CHECK;
+        PRECONDITION(CheckPointer(ppNewobjHelper));
+        PRECONDITION(CheckPointer(ppMT));
+        PRECONDITION(CheckPointer(pfFailedWhileRunningCctor));
+        PRECONDITION(*ppNewobjHelper == NULL);
+        PRECONDITION(*ppMT == NULL);
+        PRECONDITION(!*pfFailedWhileRunningCctor);
+    }
+    CONTRACTL_END;
 
-    OBJECTREF           retVal  = NULL;
-    REFLECTCLASSBASEREF objType = (REFLECTCLASSBASEREF) objTypeUNSAFE;
+    BEGIN_QCALL;
 
-    HELPER_METHOD_FRAME_BEGIN_RET_NOPOLL();
+    TypeHandle typeHandle = pTypeHandle.AsTypeHandle();
 
-    TypeHandle type = objType->GetType();
+    // Don't allow void
+    if (typeHandle.GetSignatureCorElementType() == ELEMENT_TYPE_VOID)
+    {
+        COMPlusThrow(kArgumentException, W("NotSupported_Type"));
+    }
 
-    // Don't allow void, arrays, pointers, byrefs or function pointers.
-    if (type.IsTypeDesc() || type.IsArray() || type.GetSignatureCorElementType() == ELEMENT_TYPE_VOID)
-        COMPlusThrow(kArgumentException, W("Argument_InvalidValue"));
+    // Don't allow arrays, pointers, byrefs, or function pointers
+    if (typeHandle.IsTypeDesc() || typeHandle.IsArray())
+    {
+        COMPlusThrow(fGetUninitializedObject ? kArgumentException : kMissingMethodException, W("NotSupported_Type"));
+    }
 
-    MethodTable *pMT = type.AsMethodTable();
+    MethodTable* pMT = typeHandle.AsMethodTable();
     PREFIX_ASSUME(pMT != NULL);
 
-    //We don't allow unitialized Strings or Utf8Strings.
-    if (pMT == g_pStringClass) {
-        COMPlusThrow(kArgumentException, W("Argument_NoUninitializedStrings"));
+    pMT->EnsureInstanceActive();
```

Comment:
The observable effect of this method is that it runs module constructors. Is there a reason to do this early before the argument checks?

Reply:
Currently `RuntimeTypeHandle::CreateInstance` calls this before finishing the argument checks ([see code](https://github.com/dotnet/runtime/blob/07193b828792e3e3a4754854741f7e47764b964a/src/coreclr/src/vm/reflectioninvocation.cpp#L408)), so I followed that same pattern. As long as the arg checks aren't dependent on this I can move it later in the method. Would it make sense to move it immediately before the cctor checks?

```cpp
pMT->EnsureInstanceActive(); // <-- move it to here, just before the cctor checks

// Run the type's cctor if needed (if not marked beforefieldinit)
if (pMT->HasPreciseInitCctors())
{
    *pfFailedWhileRunningCctor = TRUE;
    pMT->CheckRunClassInitAsIfConstructingThrowing();
    *pfFailedWhileRunningCctor = FALSE;
}
```

Reply:
Yes, I think so.